# Investigate a dataset project
## Introduction
This project is my implementation of the "Investigate a dataset" project. This is the second project of the Data Analyst nanodegree program of Udacity. The dataset I selected is the "No-show appointments dataset".


In [54]:
import pandas as pd
df = pd.read_csv('noshowappointments-kagglev2-may-2016.csv')
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

## A first look at the quality of the data
First we will take a look at the quality of the data using the same steps as in "Lesson 3: The data analysis proces".

In [55]:
# this returns a tuple of the dimensions of the dataframe
df.shape

(110527, 14)

In [56]:
# this returns the datatypes of the columns
df.dtypes

PatientId         float64
AppointmentID       int64
Gender             object
ScheduledDay       object
AppointmentDay     object
Age                 int64
Neighbourhood      object
Scholarship         int64
Hipertension        int64
Diabetes            int64
Alcoholism          int64
Handcap             int64
SMS_received        int64
No-show            object
dtype: object

In [57]:
# this displays a concise summary of the dataframe,
# including the number of non-null values in each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
PatientId         110527 non-null float64
AppointmentID     110527 non-null int64
Gender            110527 non-null object
ScheduledDay      110527 non-null object
AppointmentDay    110527 non-null object
Age               110527 non-null int64
Neighbourhood     110527 non-null object
Scholarship       110527 non-null int64
Hipertension      110527 non-null int64
Diabetes          110527 non-null int64
Alcoholism        110527 non-null int64
Handcap           110527 non-null int64
SMS_received      110527 non-null int64
No-show           110527 non-null object
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


All columns have 110527 non-null values which is equal to the total number of rows so we have no nulls anywhere. 

In [58]:
# this returns the number of unique values in each column
df.nunique()

PatientId          62299
AppointmentID     110527
Gender                 2
ScheduledDay      103549
AppointmentDay        27
Age                  104
Neighbourhood         81
Scholarship            2
Hipertension           2
Diabetes               2
Alcoholism             2
Handcap                5
SMS_received           2
No-show                2
dtype: int64

AppontmentID has 110527 unique values which is equal to the number of rows so we have no duplicated rows. Furthermore we see that Gender has two unique values ('F' and 'M'), No-show has two unique values ('Yes' and 'No'). There are 81 differend Neighborhoods and 5 handicap values (0 to 4). Scholarship, hipertension, diabetes, alcoholism, and SMS_received all have two possible values 0 and 1. There are 104 unique age values. That seems like a lot  

In [59]:
#List unique values in the df['name'] column
df.Age.unique()

array([ 62,  56,   8,  76,  23,  39,  21,  19,  30,  29,  22,  28,  54,
        15,  50,  40,  46,   4,  13,  65,  45,  51,  32,  12,  61,  38,
        79,  18,  63,  64,  85,  59,  55,  71,  49,  78,  31,  58,  27,
         6,   2,  11,   7,   0,   3,   1,  69,  68,  60,  67,  36,  10,
        35,  20,  26,  34,  33,  16,  42,   5,  47,  17,  41,  44,  37,
        24,  66,  77,  81,  70,  53,  75,  73,  52,  74,  43,  89,  57,
        14,   9,  48,  83,  72,  25,  80,  87,  88,  84,  82,  90,  94,
        86,  91,  98,  92,  96,  93,  95,  97, 102, 115, 100,  99,  -1])

There are some old people here, up to 115 years. There is also the value -1. A fetus? Age unknown? This is a nonsensical value.

In [60]:
df = df.drop(df.index[df.Age == -1])
df.shape

(110526, 14)

One row with Age == -1 is removed. Next are there really 81 differend neighborhoods?

In [61]:
sorted(df.Neighbourhood.unique())

['AEROPORTO',
 'ANDORINHAS',
 'ANTÔNIO HONÓRIO',
 'ARIOVALDO FAVALESSA',
 'BARRO VERMELHO',
 'BELA VISTA',
 'BENTO FERREIRA',
 'BOA VISTA',
 'BONFIM',
 'CARATOÍRA',
 'CENTRO',
 'COMDUSA',
 'CONQUISTA',
 'CONSOLAÇÃO',
 'CRUZAMENTO',
 'DA PENHA',
 'DE LOURDES',
 'DO CABRAL',
 'DO MOSCOSO',
 'DO QUADRO',
 'ENSEADA DO SUÁ',
 'ESTRELINHA',
 'FONTE GRANDE',
 'FORTE SÃO JOÃO',
 'FRADINHOS',
 'GOIABEIRAS',
 'GRANDE VITÓRIA',
 'GURIGICA',
 'HORTO',
 'ILHA DAS CAIEIRAS',
 'ILHA DE SANTA MARIA',
 'ILHA DO BOI',
 'ILHA DO FRADE',
 'ILHA DO PRÍNCIPE',
 'ILHAS OCEÂNICAS DE TRINDADE',
 'INHANGUETÁ',
 'ITARARÉ',
 'JABOUR',
 'JARDIM CAMBURI',
 'JARDIM DA PENHA',
 'JESUS DE NAZARETH',
 'JOANA D´ARC',
 'JUCUTUQUARA',
 'MARIA ORTIZ',
 'MARUÍPE',
 'MATA DA PRAIA',
 'MONTE BELO',
 'MORADA DE CAMBURI',
 'MÁRIO CYPRESTE',
 'NAZARETH',
 'NOVA PALESTINA',
 'PARQUE INDUSTRIAL',
 'PARQUE MOSCOSO',
 'PIEDADE',
 'PONTAL DE CAMBURI',
 'PRAIA DO CANTO',
 'PRAIA DO SUÁ',
 'REDENÇÃO',
 'REPÚBLICA',
 'RESISTÊNCIA',
 'RO

Apparently there are indeed 81 differend neighbourhoods. 
## Question. What are the most important factors in determing the likelyhood of a no-show?

In [62]:
#Find the overall No-show perccentage
total = len(df.index)
print('Total = ', total)
total_noshow = (df['No-show'] == 'Yes').sum()
print('General no-show = ', int((total_noshow / total) * 100), '%')

Total =  110526
General no-show =  20 %


In [63]:
#Gender
total_male = (df['Gender'] == 'M').sum()
total_female = (df['Gender'] == 'F').sum()
print(int((total_male / total) * 100), '% of the appointments is for a man')
print(int((total_female / total) * 100), '% of the appointments is for a woman')
male_noshow = len((df[(df['Gender'] == 'M') & (df['No-show'] == 'Yes')]))
female_noshow = len((df[(df['Gender'] == 'F') & (df['No-show'] == 'Yes')]))
print(int((male_noshow / total_male) * 100), '% of all appointments with a man is a no show')
print(int((female_noshow / total_female) * 100), '% of all appointments with a woman is a no show')



35 % of the appointments is for a man
64 % of the appointments is for a woman
19 % of all appointments with a man is a no show
20 % of all appointments with a woman is a no show


In [64]:
#hipertension diabetes alcoholism
total_hipertension = (df['Hipertension'] == 1).sum()
hipertension_noshow = len((df[(df['Hipertension'] == 1) & (df['No-show'] == 'Yes')]))
print(int((hipertension_noshow / total_hipertension) * 100), '% of all appointments with someone with hipertension is a no show')

total_diabetes = (df['Diabetes'] == 1).sum()
diabetes_noshow = len((df[(df['Diabetes'] == 1) & (df['No-show'] == 'Yes')]))
print(int((diabetes_noshow / total_diabetes) * 100), '% of all appointments with someone with diabetes is a no show')

total_alcoholism = (df['Alcoholism'] == 1).sum()
alcoholism_noshow = len((df[(df['Alcoholism'] == 1) & (df['No-show'] == 'Yes')]))
print(int((alcoholism_noshow / total_alcoholism) * 100), '% of all appointments with someone with alcoholism is a no show')


17 % of all appointments with someone with hipertension is a no show
18 % of all appointments with someone with diabetes is a no show
20 % of all appointments with someone with alcoholism is a no show


In [65]:
#SMS
total_sms = (df['SMS_received'] == 1).sum()
sms_noshow = len((df[(df['SMS_received'] == 1) & (df['No-show'] == 'Yes')]))
print(int((sms_noshow / total_sms) * 100), '% of all appointments with someone who received an SMS is a no show')


27 % of all appointments with someone who received an SMS is a no show


In [66]:
#In order to make the dataset less unwieldy drop some columns we no longer need
df.drop('Scholarship', axis=1, inplace=True)
df.drop('Hipertension', axis=1, inplace=True)
df.drop('Diabetes', axis=1, inplace=True)
df.drop('Alcoholism', axis=1, inplace=True)
df.drop('Handcap', axis=1, inplace=True)
df.drop('SMS_received', axis=1, inplace=True)
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood No-show  
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA      No  
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA      No  
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA      No  
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI      No  
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA      No

In [67]:
#It is easier to calculate with 1 and 0 than yes and no
df['No-show'].replace(('Yes', 'No'), (1.0, 0.0), inplace=True)

In [68]:
#age
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
sns.set_style('darkgrid')



bins = [0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100, 105, 110, 115, 120]
df['Agebins'] = pd.cut(df['Age'], bins)
df.head()

#agegrouped.head()
# print(agegrouped)

# def calculate_percent_for_age_group(min_value, max_value):
#     total_age = len((df[(df['Age'] >= min_value) & (df['Age'] < max_value)]))
#     noshow_age = len((df[(df['Age'] >= min_value) & (df['Age'] < max_value) & (df['No-show'] == 'Yes')]))
#     print(total_age)
#     if total_age < 1:
#         return None
#     else:
#         return int((noshow_age / total_age) * 100)

# age_values_min = np.arange(0, 120, 5)

# for x in np.nditer(age_values_min):
#     result = calculate_percent_for_age_group(x, x + 4)
#     if result != None:
#         print(x, x + 4, result)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  No-show   Agebins  
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA      0.0  (60, 65]  
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA      0.0  (55, 60]  
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA      0.0  (60, 65]  
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI      0.0   (5, 10]  
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA      0.0  (55, 60]

In [69]:
df.groupby('Agebins')

In [70]:
#neighbourhood

# def calculate_percent_noshow_for_neighbourhood(neighbourhood):
#     total_neighbourhood = len(df['Neighbourhood'] == neighbourhood)
#     noshow_neighbourhood = len((df[(df['Neighbourhood'] == neighbourhood) & (df['No-show'] == 'Yes')]))
#     return int((noshow_neighbourhood / total_neighbourhood) * 100)
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  No-show   Agebins  
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA      0.0  (60, 65]  
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA      0.0  (55, 60]  
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA      0.0  (60, 65]  
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI      0.0   (5, 10]  
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA      0.0  (55, 60]

In [71]:
#date between scheduled and appointment

#Drop the time part
df['ScheduledDay'] = df['ScheduledDay'].str.slice(start=0, stop=10)
df['AppointmentDay'] = df['AppointmentDay'].str.slice(start=0, stop=10)

#convert to datetime and subtract to get the days
df['ScheduledDay'] = pd.to_datetime(df['ScheduledDay'])
df['AppointmentDay'] = pd.to_datetime(df['AppointmentDay'])
df['DaysBetween'] = df['AppointmentDay'] - df['ScheduledDay']
df['DaysBetween'] = df['DaysBetween'].dt.days

df['DayOfWeek'] = df['AppointmentDay'].dt.weekday_name
df.head()
#df.tail()


PatientId  AppointmentID Gender ScheduledDay AppointmentDay  Age  \
0  2.987250e+13        5642903      F   2016-04-29     2016-04-29   62   
1  5.589978e+14        5642503      M   2016-04-29     2016-04-29   56   
2  4.262962e+12        5642549      F   2016-04-29     2016-04-29   62   
3  8.679512e+11        5642828      F   2016-04-29     2016-04-29    8   
4  8.841186e+12        5642494      F   2016-04-29     2016-04-29   56   

       Neighbourhood  No-show   Agebins  DaysBetween DayOfWeek  
0    JARDIM DA PENHA      0.0  (60, 65]            0    Friday  
1    JARDIM DA PENHA      0.0  (55, 60]            0    Friday  
2      MATA DA PRAIA      0.0  (60, 65]            0    Friday  
3  PONTAL DE CAMBURI      0.0   (5, 10]            0    Friday  
4    JARDIM DA PENHA      0.0  (55, 60]            0    Friday

In [72]:
#day of week
#type(df['DaysBetween'][0])
#df['DaysBetween'].max
print(df.groupby('Neighbourhood').mean()['No-show'].max()
# type(df2)

SyntaxError: unexpected EOF while parsing (<ipython-input-72-285e653a7849>, line 5)